data/pools/transactionPool.go

```diff
@@ -295,35 +233,137 @@ func (pool *TransactionPool) Remember(txgroup []transactions.SignedTxn) error {
 	pool.mu.Lock()
 	defer pool.mu.Unlock()
 
-	err := pool.test(txgroup)
-	if err != nil {
-		return fmt.Errorf("TransactionPool.Remember: %v", err)
+	return pool.test(txgroup)
+}
+
+type poolIngestParams struct {
+	remember bool // if set, remember txn
+	checkFee bool // if set, perform fee checks
+	mustSync bool // if set, wait until ledger is caught up
+}
+
+// remember attempts to add a transaction group to the pool.
+func (pool *TransactionPool) remember(txgroup []transactions.SignedTxn) error {
+	params := poolIngestParams{
+		remember: true,
+		checkFee: true,
+		mustSync: true,
+	}
+	return pool.ingest(txgroup, params)
+}
+
+// add tries to add the transaction group to the pool, bypassing the fee
+// priority checks.
+func (pool *TransactionPool) add(txgroup []transactions.SignedTxn) error {
+	params := poolIngestParams{
+		remember: true,
+		checkFee: false,
+		mustSync: false,
 	}
+	return pool.ingest(txgroup, params)
+}
 
+// test checks whether a transaction group could be remembered in the pool,
+// but does not actually store this transaction in the pool.
+func (pool *TransactionPool) test(txgroup []transactions.SignedTxn) error {
+	params := poolIngestParams{
+		remember: false,
+		checkFee: true,
+		mustSync: true,
+	}
+	return pool.ingest(txgroup, params)
+}
+
+// ingest checks whether a transaction group could be remembered in the pool,
+// and stores this transaction in the pool if remember is set.
+//
+// ingest assumes that pool.mu is locked.  It might release the lock
+// while it waits for OnNewBlock() to be called.
+func (pool *TransactionPool) ingest(txgroup []transactions.SignedTxn, params poolIngestParams) error {
 	if pool.pendingBlockEvaluator == nil {
-		return fmt.Errorf("TransactionPool.Remember: no pending block evaluator")
+		return fmt.Errorf("TransactionPool.test: no pending block evaluator")
 	}
 
-	err = pool.remember(txgroup)
-	if err != nil {
-		return fmt.Errorf("TransactionPool.Remember: %v", err)
+	if params.mustSync {
+		// Make sure that the latest block has been processed by OnNewBlock().
+		// If not, we might be in a race, so wait a little bit for OnNewBlock()
+		// to catch up to the ledger.
+		latest := pool.ledger.Latest()
+		waitExpires := time.Now().Add(timeoutOnNewBlock)
+		for pool.pendingBlockEvaluator.Round() <= latest && time.Now().Before(waitExpires) {
+			condvar.TimedWait(&pool.cond, timeoutOnNewBlock)
+			if pool.pendingBlockEvaluator == nil {
+				return fmt.Errorf("TransactionPool.test: no pending block evaluator")
+			}
+		}
+	}
+
+	if params.checkFee {
+		err := pool.checkSufficientFee(txgroup)
+		if err != nil {
+			return err
+		}
+	}
+
+	if !params.remember {
+		tentativeRound := pool.pendingBlockEvaluator.Round() + pool.numPendingWholeBlocks
+		err := pool.pendingBlockEvaluator.TestTransactionGroup(txgroup)
+		if err == ledger.ErrNoSpace {
+			tentativeRound++
```

Comment:
I think you're missing retry logic here

Reply:
It looks like the old code is also missing retry logic? https://github.com/algorand/go-algorand/pull/525/files/cb8efe1d421fc22aec7b2475b35b00ef28174bee#diff-d61c312f7cb0cc88e1721926edafe92bL220

Reply:
Indeed, looks like you are correct.  Maybe it was correct before we added txgroups, but became incorrect after txgroups were added?  In any case, yeah, looks like the current test code could return `true` when the pending block evaluator runs out of space for a part of the group's transactions, even if later transactions in the group are invalid..

Reply:
I think if we try to fix this by calling `ResetTxnBytes`, `test` will no longer be read-only and will possibly cause us to prematurely increment the next round to be checked by the pool's block evaluator.

Reply:
Perhaps I'll keep the buggy behavior for now--I think in the worst case, we do extra crypto work on a transaction group. I guess the right behavior would be to save the old `blockTxBytes` in the evaluator, test, and then restore it after the call?

Reply:
One possible fix would be to change the API for `TestTransactionGroup` to allow the caller to inflate the max block size limit for the duration of that `TestTransactionGroup` call.

Reply:
Chatted with Max--the only reason we want `Test` is to do duplicate detection before doing any crypto. We don't need `Test` to be secure against an attacker since any attacker can always give us a bad signature and force us to do crypto. I think I'll modify `Test` to do much less--just check for duplicates and well-formedness.

Reply:
Trimming down `Test` sounds like a great plan!